src/blockrelay/blockrelay_common.cpp

```diff
@@ -318,30 +316,51 @@ std::shared_ptr<CBlockThinRelay> ThinTypeRelay::SetBlockToReconstruct(CNode *pfr
         pblock->cmpctblock = std::make_shared<CompactBlock>(CompactBlock());
         pblock->grapheneblock = std::make_shared<CGrapheneBlock>(CGrapheneBlock());
 
-        mapBlocksReconstruct.insert(
-            std::make_pair(pfrom->GetId(), std::make_pair(pblock->GetBlockHeader().GetHash(), pblock)));
+        mapBlocksReconstruct.insert(std::make_pair(pfrom->GetId(), std::make_pair(hash, pblock)));
         return pblock;
     }
 }
 
-std::shared_ptr<CBlockThinRelay> ThinTypeRelay::GetBlockToReconstruct(CNode *pfrom)
+std::shared_ptr<CBlockThinRelay> ThinTypeRelay::GetBlockToReconstruct(CNode *pfrom, const uint256 &hash)
 {
     // Retrieve a current instance of a block being reconstructed. This is typically used
     // when we have received the response of a re-request for more transactions.
     LOCK(cs_reconstruct);
-    auto iter = mapBlocksReconstruct.find(pfrom->GetId());
-    if (iter != mapBlocksReconstruct.end())
+    auto range = mapBlocksReconstruct.equal_range(pfrom->GetId());
+    while (range.first != range.second)
     {
-        return iter->second.second;
+        if (range.first->second.first == hash)
+        {
+            return range.first->second.second;
+        }
+        range.first++;
     }
-    else
-        return nullptr;
+    return nullptr;
 }
 
-void ThinTypeRelay::ClearBlockToReconstruct(NodeId id)
+void ThinTypeRelay::ClearBlockToReconstruct(NodeId id, const uint256 &hash)
 {
     LOCK(cs_reconstruct);
-    mapBlocksReconstruct.erase(id);
+    auto range = mapBlocksReconstruct.equal_range(id);
+    while (range.first != range.second)
+    {
+        if (range.first->second.first == hash)
+        {
+            mapBlocksReconstruct.erase(range.first);
+            break;
+        }
+        range.first++;
+    }
+}
+
+void ThinTypeRelay::ClearAllBlocksToReconstruct(NodeId id)
+{
+    LOCK(cs_reconstruct);
+    auto range = mapBlocksReconstruct.equal_range(id);
+    while (range.first != range.second)
```

Comment:
use one of these APIs to avoid the while loop
(http://www.cplusplus.com/reference/map/multimap/erase/)
size_type erase (const key_type& k);
void erase (iterator first, iterator last);

Reply:
done